#generate.length.distributions.R
#plots all genome-matching sRNA-Seq read length distributions as shown in Supplementary Figure 2A-2C

#install.packages("extrafont")
library("extrafont")
#font_import()
loadfonts()

setwd('/Volumes/nodine/lab/members/michael/compFiles/sRNAs/projects/apps/smallRNA_spikeIns/scripts.for.ms/')  #user will have to change this to their working directory

sample1 = 'col0_fb1' #biorep 1; could be either col0_leaf1, col0_fb1 or d234_fb1 for panels a, b and c, respectively
sample2 = 'col0_fb2' #biorep 1; could be either col0_leaf2, col0_fb2 or d234_fb2 for panels a, b and c, respectively
sample_all = 'd234_fb' #set name; could be either col0_leaf, col0_fb or d234_fb for panels a, b and c, respectively
start = 18 #smallest sRNA-Seq read size to be considered
end = 31 #largest sRNA-Seq read size (+1) to be considered
geneType = 'all' #considers all genome-matching reads 
pos = 0   #position within read, in this case the first base, to look at base frequencies in stacked bar chart
fileName = paste('lenDistrib__geneType_',geneType,'__start_',start,'__end_',end,'__pos_',pos,sep='')  #name of input file generated by getLengthVals.py
ymax = 325 #maximum value on y-axis of graph

options("scipen"=1)

#print length distributions all genome-matching reads as rpm with base preferences of first position indicated
ntMatrix_1 = read.table(paste('methods.data.analysis/small.RNA.Seq/',sample1,'/lenDistrib/',fileName,sep=''), header=T, sep="\t", row.names=1)
ntMatrix_2 = read.table(paste('methods.data.analysis/small.RNA.Seq/',sample2,'/lenDistrib/',fileName,sep=''), header=T, sep="\t", row.names=1)

#divide by 1000 for plotting purposes
ntMatrix_1 = ntMatrix_1/1000
ntMatrix_2 = ntMatrix_2/1000

ntMatrix_1 = t(ntMatrix_1)
ntMatrix_2 = t(ntMatrix_2)

ntMatrix_all = (ntMatrix_1 + ntMatrix_2)/2

ntMatrix_1 = as.matrix(ntMatrix_1)
ntMatrix_2 = as.matrix(ntMatrix_2)

ntMatrix_all = as.matrix(ntMatrix_all)

col_v = c('green','red','blue','yellow')

main_lab = paste(sample_all,'_',geneType,sep='')

plotFile = paste("supplemental.figure.2/",sample_all,".length.distribution.pdf",sep='')
pdf(plotFile, family="Arial", useDingbats=FALSE)
mp = barplot(height=ntMatrix_all, col = col_v, main=main_lab, xlab="Length (nucleotides)", ylab="Reads per million (in thousands)", las=1, ylim=c(0,ymax))
legend_coords = xy.coords(mp[12],ymax)
legend(legend_coords, row.names(ntMatrix_1), fill=col_v, box.lty=0)
#add error bars
group1 = c(sum(ntMatrix_1[,1]),sum(ntMatrix_2[,1]))
group2 = c(sum(ntMatrix_1[,2]),sum(ntMatrix_2[,2]))
group3 = c(sum(ntMatrix_1[,3]),sum(ntMatrix_2[,3]))
group4 = c(sum(ntMatrix_1[,4]),sum(ntMatrix_2[,4]))
group5 = c(sum(ntMatrix_1[,5]),sum(ntMatrix_2[,5]))
group6 = c(sum(ntMatrix_1[,6]),sum(ntMatrix_2[,6]))
group7 = c(sum(ntMatrix_1[,7]),sum(ntMatrix_2[,7]))
group8 = c(sum(ntMatrix_1[,8]),sum(ntMatrix_2[,8]))
group9 = c(sum(ntMatrix_1[,9]),sum(ntMatrix_2[,9]))
group10 = c(sum(ntMatrix_1[,10]),sum(ntMatrix_2[,10]))
group11 = c(sum(ntMatrix_1[,11]),sum(ntMatrix_2[,11]))
group12 = c(sum(ntMatrix_1[,12]),sum(ntMatrix_2[,12]))
group13 = c(sum(ntMatrix_1[,13]),sum(ntMatrix_2[,13]))

repSize = 3

arrows(mp[1], mean(group1) + sd(group1)/sqrt(repSize),mp[1],mean(group1) - sd(group1)/sqrt(repSize), angle=90, code=3, length=0.1)
arrows(mp[2], mean(group2) + sd(group2)/sqrt(repSize),mp[2],mean(group2) - sd(group2)/sqrt(repSize), angle=90, code=3, length=0.1)
arrows(mp[3], mean(group3) + sd(group3)/sqrt(repSize),mp[3],mean(group3) - sd(group3)/sqrt(repSize), angle=90, code=3, length=0.1)
arrows(mp[4], mean(group4) + sd(group4)/sqrt(repSize),mp[4],mean(group4) - sd(group4)/sqrt(repSize), angle=90, code=3, length=0.1)
arrows(mp[5], mean(group5) + sd(group5)/sqrt(repSize),mp[5],mean(group5) - sd(group5)/sqrt(repSize), angle=90, code=3, length=0.1)
arrows(mp[6], mean(group6) + sd(group6)/sqrt(repSize),mp[6],mean(group6) - sd(group6)/sqrt(repSize), angle=90, code=3, length=0.1)
arrows(mp[7], mean(group7) + sd(group7)/sqrt(repSize),mp[7],mean(group7) - sd(group7)/sqrt(repSize), angle=90, code=3, length=0.1)
arrows(mp[8], mean(group8) + sd(group8)/sqrt(repSize),mp[8],mean(group8) - sd(group8)/sqrt(repSize), angle=90, code=3, length=0.1)
arrows(mp[9], mean(group9) + sd(group9)/sqrt(repSize),mp[9],mean(group9) - sd(group9)/sqrt(repSize), angle=90, code=3, length=0.1)
arrows(mp[10], mean(group10) + sd(group10)/sqrt(repSize),mp[10],mean(group10) - sd(group10)/sqrt(repSize), angle=90, code=3, length=0.1)
arrows(mp[11], mean(group11) + sd(group11)/sqrt(repSize),mp[11],mean(group11) - sd(group11)/sqrt(repSize), angle=90, code=3, length=0.1)
arrows(mp[12], mean(group12) + sd(group12)/sqrt(repSize),mp[12],mean(group12) - sd(group12)/sqrt(repSize), angle=90, code=3, length=0.1)
arrows(mp[13], mean(group13) + sd(group13)/sqrt(repSize),mp[13],mean(group13) - sd(group13)/sqrt(repSize), angle=90, code=3, length=0.1)


dev.off()
